src/pip/_internal/models/link.py

```diff
@@ -372,6 +372,11 @@ def url_without_fragment(self) -> str:
         scheme, netloc, path, query, fragment = self._parsed_url
         return urllib.parse.urlunsplit((scheme, netloc, path, query, ""))
 
+    @property
+    def metadata_file_url(self) -> str:
+        scheme, netloc, path, query, fragment = self._parsed_url
+        return urllib.parse.urlunsplit((scheme, netloc, path + ".metadata", query, ""))
```

Comment:
Do we want to keep the query? In #12033 the query is used to implement Base64 checking, and including the query here might throw off the server. The spec doesn’t really comment about this, but since the original query is for the original URL, it’s reasonable to _not_ include it for the new path.

Reply:
I have no idea. It's just as plausible that the query could contain an access token that allows you to fetch data from the server.

I think the only way we'd know for sure what the correct behaviour is would be to survey what existing index server implementations do. And if there's none currently using the query segment, I'd be tempted to amend the PEP to prohibit the query segment in index URLs altogether.